### PR TITLE
Wiki lookup plugin

### PR DESCRIPTION
**Pull request by Reasel**
# Lookup Items, NPCs, Objects, Quests, Spells, Diaries, and more via menu option.

This plugin will allow you to add a new option to Lookup which when clicked will open the  **new** wiki page for that thing in your default browser.

It functions only when holding shift. It works in the bank, on the ground, in your inventory, and even your equipment.

Things you can wiki:

- Inventory Items
- Equipment
- Skills
- Spells
- Prayers
- Music
- Quests
- Diaries
- Objects

# Examples
This
settings1
Will produce this
allOff

This
settings2
Will produce this. As long as the third option is select this will be the same behavior regardless of the other settings.
alwaysOn

This
settings3
Will Produce this (Also to add here is that if you click cancel on anything it will turn off the wiki search)
DefaultNoShift

This
settings4
Will produce this
shiftPlusOrbDeafault

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupInput.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.plugins.wikilookup;
+
+import net.runelite.client.input.KeyListener;
+
+import javax.inject.Inject;
+import java.awt.event.KeyEvent;
+
+public class WikiLookupInput implements KeyListener
```
Comment (AlexKolpa) [inline runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupInput.java]: I feel like an inner class might be more suited, or at least make it package-private. The coupling between the plugin and this listener seems a bit too tight to have exposed as a public class.

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupPlugin.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+				addBlankOption("cancel", WIKI, event.getOption().replaceFirst("View ", "").replaceFirst(" guide", ""));
+			}
+		}
+	}
+
+	private void changeOption(int actionID, String newOption)
+	{
+		MenuEntry[] entries = client.getMenuEntries();
+
+		for (int i = 0; i < entries.length; i++)
```
Comment (AlexKolpa) [inline runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupPlugin.java]: Minor (same applies to `changeTarget`): Since you don't rely on the index of the element, you could also do a for-each loop here.

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupPlugin.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+	private void addBlankOption(String toCopy, String newOption, String newTarget)
+	{
+		MenuEntry[] entries = client.getMenuEntries();
+		MenuEntry[] newEntries = new MenuEntry[entries.length + 1];
+		for (int i = 0; i < entries.length; ++i)
+		{
+			if (Text.removeTags(entries[i].getOption()).toLowerCase().contains(toCopy))
+			{
+				MenuEntry toInsert = new MenuEntry();
```
Comment (AlexKolpa) [inline runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupPlugin.java]: What's the reason you dropped the `toBuilder` approach here?

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupInput.java (hunk 1 above)
Comment (Reasel) [reply runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupInput.java]: I think you may be correct.

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupPlugin.java (hunk 3 above)
Comment (Reasel) [reply runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupPlugin.java]: I dropped it because there was a conflict with the way the code functions in other places in bytecode. I am not wholly sure but at one point I had changed every instance in the codebase but was getting an error at a certain line but could not edit it. 

This way doesn't change anything else and only is messy in this one spot. Someone with more knowledge should look into making the toBuilder() approach.

Comment (AlexKolpa) [reply runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupPlugin.java]: Alright, clear. It's supposed to be a useful tool, and thus shouldn't be causing annoyances like this. Indeed probably not worth it right now, at least not for this PR.

Comment (Reasel) [reply runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupPlugin.java]: Yeah, I did have it all working. It is really nice, but there was a few places that I could not alter that needed the default constructor method. Honestly I think it was how rs did it? I am not sure.

Comment (AlexKolpa) [reply runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupPlugin.java]: Couldn't you just allow both `MenuEntry.builder()` and `new MenuEntry()`?

Comment (Reasel) [reply runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupPlugin.java]: I have found a way to do just this.

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupPlugin.java (hunk 2 above)
Comment (Reasel) [reply runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupPlugin.java]: In the if statement we are using the index to reference that entry `entries[i].setOption(newOption);` Using a foreach would be the same efficiency as well as make the code messier.

Comment (AlexKolpa) [reply runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupPlugin.java]: Definitely agree that from an efficiency perspective it doesn't really matter, but just pointing out it might improve readability here.

**Comment by Reasel** [comment]
Adding link to previous PR's [discussion](https://github.com/runelite/runelite/pull/2304).

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupPlugin.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+	private void addBlankOption(String toCopy, String newOption, String newTarget)
+	{
+		MenuEntry[] entries = client.getMenuEntries();
+		MenuEntry[] newEntries = new MenuEntry[entries.length + 1];
+		for (int i = 0; i < entries.length; ++i)
+		{
+			if (Text.removeTags(entries[i].getOption()).toLowerCase().contains(toCopy))
+			{
+				MenuEntry toInsert = MenuEntry.builder()
```
[the remark this answers is not in the document]
Comment (Reasel) [reply runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupPlugin.java]: I believe you are correct. Fixed in most recent commit.

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupInput.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.plugins.wikilookup;
+
+import net.runelite.client.input.KeyListener;
+
+import javax.inject.Inject;
+import java.awt.event.KeyEvent;
+
+class WikiLookupInput implements KeyListener
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupInput.java]: You can probably implement this interface directly into plugin to reduce the need for getters and setters etcetc

**Comment by Nightfirecat** [comment]
What's the status on this PR?

**Comment by Reasel** [comment]
I haven't been working on this in awhile. I will take a look at this since it seems simple to change. @deathbeam do you have an example of another plugin that handles their input listener how you described?

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupPlugin.java [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+	private boolean hotKeyPressed = false;
+
+	@Override
+	public void keyTyped(KeyEvent e)
+	{
+
+	}
+
+	@Override
+	public void keyPressed(KeyEvent e)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupPlugin.java]: We now even have Hotkey class that you can use instead of this

Comment (Reasel) [reply runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupPlugin.java]: Well then I guess I have some work to do!

Comment (Reasel) [reply runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupPlugin.java]: I think we do not want to use that since I am trying to detect when the shift key is held and a menu is opened NOT just when the hotkey is pressed like in the Screenshot plugin.

**Comment by Reasel** [comment]
@Nightfirecat the status is now waiting for merging.

**Comment by Nightfirecat** [comment]
Looks like this is conflicting with HiscorePlugin x_x

**Comment by Reasel** [comment]
Should be an easy fix.

**Comment by Nightfirecat** [comment]
This feature should probably wait until this migration happens: https://runescape.wikia.com/wiki/Forum:Leaving_Wikia

tl;dr version: rs and osrs wikis are moving domains and leaving Wikia, to be co-owned by Jagex

**Comment by Reasel** [comment]
@Nightfirecat Yeah hopefully the new home is workable.

**Comment by deathbeam** [comment]
So I just randomly saw this reddit thread: https://www.reddit.com/r/2007scape/comments/9j5f5l/ingame_wiki_lookup_orb_tool_here_you_go_mod_sween/ and actually, having button like that next to world map is way cleaner than having it as shift keybinding imo

**Comment by Reasel** [comment]
That's fine, I can work on that. I just think id rather have an option for a key bind since sometimes you can't click on the exact thing. But some visual thing would be best for some people for sure.

**Comment by MagicfTail** [comment]
https://i.imgur.com/IiYiZpl.mp4 Someone on Reddit made a gif of how it could look in game

**Comment by MagicfTail** [comment]
https://www.reddit.com/r/2007scape/comments/9jnwcw/wiki_orb/ and the post itself, lots of feedback

**Comment by Reasel** [comment]
@MagicfTail Yeah I saw that. I will look into adding something like that, but this PR is simple enough that anyone should be able to change it. Most of the logic is tied up in formatting the different menu entries in order to build the URL. Any sort of wiki orb could use the exact same thing since the menu entries are built on hover and only the first option is shown.

**Comment by deathbeam** [comment]
Imo the orb is better solution than keybind, since basically what would you do when pressing the orb is toggle the menu entry, and make it first and on click you just disable the menu entry adding. So it will work in all places where keybind would but would be nicer to use

**Comment by Reasel** [comment]
My only concern with the orb is that you will only be able to wiki the top most item of a stack. Works great for everything in your inventory and interfaces, but here is an example of something it would not work well for: 

Stack

**Comment by Reasel** [comment]
So for anyone who was waiting for this PR. I think all it needs now is polish and feedback from you all.

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupConfig.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+@ConfigGroup("wikilookup")
+public interface WikiLookupConfig extends Config
+{
+	@ConfigItem(
+			position = 0,
+			keyName = "boundtoshift",
+			name = "Shift bound lookup",
+			description = "Determines if wiki lookup will enable while shift is held."
+	)
+	default boolean boundtoshift()
```
Comment (raiyni) [inline runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupConfig.java]: camelCase your config methods

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupConfig.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+		return false;
+	}
+	
+	@ConfigItem(
+			position = 2,
+			keyName = "permwiki",
+			name = "Make wiki option always visible",
+			description = "Makes the wiki option always visible but never the default regardless of options."
+	)
+	default boolean permwiki()
```
Comment (raiyni) [inline runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupConfig.java]: camelCase your config methods

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiOrbInterface.java [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+	{
+		if (isHidden())
+		{
+			return;
+		}
+		loaded = true;
+		parent = client.getWidget(WidgetInfo.MINIMAP_ORBS);
+		wikiOrb = createGraphic("", SpriteID.MINIMAP_MAP_ORB_FRAME, -1, 30 , 30 , 170, 0, true);
+		orbBackground = createGraphic("", SpriteID.MINIMAP_ORB_RUN, -1, 25 , 25 , 173, 3, false);
+		orbBackgroundActivated = createGraphic("", SpriteID.MINIMAP_ORB_RUN_ACTIVATED, -1, 25 , 25 , 173, 3, false);
```
Comment (raiyni) [inline runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiOrbInterface.java]: You can just set the `spriteId` instead of using a whole new widget.

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiOrbInterface.java [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+import net.runelite.api.widgets.WidgetInfo;
+import net.runelite.client.callback.ClientThread;
+import net.runelite.client.config.ConfigManager;
+
+@Slf4j
+@Singleton
+public class WikiOrbInterface
+{
+	private static final String WIKI_ORB = "Wiki Lookup";
+	private static final int TAB_HEIGHT = 40;
```
Comment (raiyni) [inline runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiOrbInterface.java]: Unused variables

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiOrbInterface.java [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+	private static final int BUTTON_HEIGHT = 20;
+	private static final int MARGIN = 1;
+	private final Client client;
+	private final ClientThread clientThread;
+	private final ConfigManager configManager;
+	private final WikiLookupConfig config;
+	private final Rectangle bounds = new Rectangle();
```
Comment (raiyni) [inline runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiOrbInterface.java]: unused

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiOrbInterface.java [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+		{
+			return;
+		}
+		loaded = true;
+		parent = client.getWidget(WidgetInfo.MINIMAP_ORBS);
+		wikiOrb = createGraphic("", SpriteID.MINIMAP_MAP_ORB_FRAME, -1, 30 , 30 , 170, 0, true);
+		orbBackground = createGraphic("", SpriteID.MINIMAP_ORB_RUN, -1, 25 , 25 , 173, 3, false);
+		orbBackgroundActivated = createGraphic("", SpriteID.MINIMAP_ORB_RUN_ACTIVATED, -1, 25 , 25 , 173, 3, false);
+		orbBackgroundActivated.setHidden(true);
+		orbFrame = createGraphic("", SpriteID.BANK_SEARCH, -1, 20, 20, 174, 3, false);
```
Comment (raiyni) [inline runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiOrbInterface.java]: Your sprite icon looks squished in the photo. I would also make these numbers less magic and calculate based on parent width/height.

Comment (Reasel) [reply runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiOrbInterface.java]: I have little to no experience in that regard. Mind helping me? Or showing me where its been done before so I can glean?

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiOrbInterface.java (hunk 11 above)
Comment (Reasel) [reply runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiOrbInterface.java]: Fixed now.

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiOrbInterface.java (hunk 10 above)
Comment (Reasel) [reply runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiOrbInterface.java]: Fixed now.

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiOrbInterface.java (hunk 12 above)
Comment (raiyni) [reply runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiOrbInterface.java]: Use the widget inspector and change the size until it looks good

**Comment by Nightfirecat** [comment]
History looks wrong? Also, I think the expected behavior (based on that reddit suggestion) was to change the left-click option to "Wiki" when the wiki orb was selected, not to add a right-click option.

**Comment by Reasel** [comment]
Oh wow, I messed something up for sure.

But I have option in the settings to change how it functions. We can set the default to whatever it's obviously now set in stone.

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupConfig.java [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+@ConfigGroup("wikilookup")
+public interface WikiLookupConfig extends Config
+{
+	@ConfigItem(
+			position = 0,
+			keyName = "boundToShift",
+			name = "Shift bound lookup",
+			description = "Determines if wiki lookup will enable while shift is held."
+	)
+	default boolean boundToShift()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupConfig.java]: This should be disabled by default

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupConfig.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+		return false;
+	}
+	
+	@ConfigItem(
+			position = 2,
+			keyName = "permWiki",
+			name = "Make wiki option always visible",
+			description = "Makes the wiki option always visible but never the default regardless of options."
+	)
+	default boolean permWiki()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupConfig.java]: How is this useful? Probs not necessary

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupConfig.java [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+		return true;
+	}
+	
+	@ConfigItem(
+			position = 1,
+			keyName = "removeNonVital",
+			name = "Make wiki option default",
+			description = "Determines whether wiki is the only option available during lookup."
+	)
+	default boolean removeNonVital()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupConfig.java]: This should be probably on by default, or at least default behaviour with the orb thing

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupConfig.java [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+	{
+		return false;
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupConfig.java]: I think this should be `true`. Most people will not expect to need to right-click after clicking the wiki orb.

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupConfig.java (hunk 14 above)
Comment (Reasel) [reply runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupConfig.java]: This is the option that simply overrides everything and makes it so that anywhere in game the wiki option is always visible but not the default....kind of like examine. Personally this is what I have been using.

File: runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupPlugin.java [hunk 17]
```diff
@@ -0,0 +1,290 @@
...
+import net.runelite.client.util.LinkBrowser;
+import net.runelite.client.util.Text;
+import org.apache.commons.lang3.ArrayUtils;
+
+import javax.inject.Inject;
+import java.awt.event.KeyEvent;
+
+@PluginDescriptor(
+		name = "Wiki Lookup",
+		enabledByDefault = false
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/wikilookup/WikiLookupPlugin.java]: I would not mind having this enabled by default, idk what others thing

**Comment by deathbeam** [comment]
There are only few problems with this PR atm:
1. The stupid click-through minimap problem, in fixed mode with this wiki orb, you click minimap like all the time what makes it unusable.
2. You need to click ground twice in order to walk

**Comment by Nightfirecat** [comment]
Could the click action be consumed if the wiki orb is clicked in fixed mode to avoid the accidental walking issue?

**Comment by jaydenkieran** [comment]
@Nightfirecat @Reasel Hi, I'm an admin on the RS & OSRS wikis, let us know if you need any help working with our new sites, but for the most part it should just be a simple URL change from `oldschoolrunescape.wikia.com/wiki/<thing>` to `oldschool.runescape.wiki/w/<thing>`.

**Comment by deathbeam** [comment]
I think I tried consuming it @Nightfirecat and it simply didnt worked

**Comment by Reasel** [comment]
@jaydenkieran I believe I have updated the plugin to use the new wiki url. Thanks for the support!

**Comment by Nightfirecat** [comment]
I think `event.consume()` doesn't work because there's not actually a menu option being clicked for the minimap... I'm not really sure of a good way to avoid this since it seems to be an issue with the Jagex orbs too. (there's a small area where "Toggle run" overlaps with the world map)

There's two things I can recommend to mitigate that issue, both inspired by the world map orb:

1. Move the wiki orb further to the top-right, farther away from the world map
2. Resize the clickable area for the orb to be smaller. You can see below that the clickbox for the world map orb is actually rather small&mdash;notably smaller than the full orb size&mdash;only the size of the bounding box drawn here.

untitled-1
